File: Combinations/Solution/combinations.java
package Combinations.Solution;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class combinations {

    public static void printCombinations(int n, List<Integer> path) {
        if (n == 0) {
            System.out.println(String.join("+", path.stream().map(Object::toString).toArray(String[]::new)));
            return;
        }
        if (n < 0) {
            return;
        }

        for (int step : new int[]{1, 2, 3}) {
            List<Integer> newPath = new ArrayList<>(path);
            newPath.add(step);
            printCombinations(n - step, newPath);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int nInput = scanner.nextInt();
        printCombinations(nInput, new ArrayList<>());
        scanner.close();
    }
}

File: Combinations_Q-8/Solution/combinations.py
def print_combinations(n, path=[]):
    if n == 0:
        print('+'.join(map(str, path)))
        return
    if n < 0:
        return

    for step in [1, 2, 3]:
        print_combinations(n - step, path + [step])

def main():
    nInput = int(input())
    print_combinations(nInput)
    
main()